java/arcs/core/allocator/Arc.kt

```diff
@@ -147,8 +147,15 @@ class Arc internal constructor(
         if (arcState == state) return this
 
         val deferred: CompletableDeferred<Arc> = CompletableDeferred()
-        onArcStateChangeFiltered(state) {
-            deferred.complete(this@Arc)
+
+        onArcStateChange {
```

Comment:
Will this listener ever go away? If not, each call to `waitFor` will add another one, which is undesirable, we should consider how to avoid/reuse?

Reply:
Yes, I think this is a good idea.

This is somewhat mitigated by the fact that the handler is only added is the arc is not already in that state. So for the bulk of the `waitFor` calls, a handler will not be added. 

But it would still be nice to make these one-shot handlers. I can file a ticket for this, if there isn't one already.

Reply:
(reminded me that we want to file a similar task for storage proxy listeners, as well!)

Reply:
We could add remove the one shot after deferred.await(), but can do it in a follow up.

Reply:
I decided to give it a shot in this PR, seems straightforward enough but please check my work :-)

Reply:
I decided to give it a shot in this PR, seems straightforward enough but please check my work :-)